File: backend/app/api/v1/endpoints/finance.py
"""
AGRIOS — Finance Module API Endpoints

Route map:
  --- Expense Categories ---
  GET    /farms/{farm_id}/finance/categories            → list categories (system + custom)
  POST   /farms/{farm_id}/finance/categories            → create custom category

  --- Expenses ---
  POST   /farms/{farm_id}/expenses                      → log expense
  GET    /farms/{farm_id}/expenses                      → list expenses (paginated)
  GET    /farms/{farm_id}/expenses/{expense_id}         → single expense
  PATCH  /farms/{farm_id}/expenses/{expense_id}         → correct expense
  DELETE /farms/{farm_id}/expenses/{expense_id}         → soft-delete

  --- Revenue ---
  POST   /farms/{farm_id}/revenue                       → log revenue
  GET    /farms/{farm_id}/revenue                       → list revenue (paginated)
  GET    /farms/{farm_id}/revenue/{record_id}           → single revenue record
  PATCH  /farms/{farm_id}/revenue/{record_id}           → correct revenue record
  DELETE /farms/{farm_id}/revenue/{record_id}           → soft-delete

  --- Financial Snapshot ---
  GET    /farms/{farm_id}/flocks/{flock_id}/finance     → flock P&L snapshot
  POST   /farms/{farm_id}/flocks/{flock_id}/finance/refresh  → force recompute snapshot

  --- Dashboard ---
  GET    /farms/{farm_id}/finance                       → farm finance dashboard
  GET    /farms/{farm_id}/finance/categories/breakdown  → expense category breakdown

  --- Calculators (pure, no DB) ---
  POST   /calculators/fcr                               → FCR calculator
  POST   /calculators/profit-projection                 → profit projection
  POST   /calculators/break-even                        → break-even price
  POST   /calculators/feed-needs                        → feed needs estimate

RBAC:
  FINANCE_RECORD → farm_owner, farm_manager
  FINANCE_VIEW   → all roles (farm_owner, farm_manager, farm_worker, vet_consultant, viewer)
"""

from datetime import date
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status

from app.core.permissions import Permission, require_permission
from app.dependencies import CurrentUser, DBSession, require_farm_access
from app.schemas.base import SuccessResponse
from app.schemas.finance import (
    BreakEvenInput,
    BreakEvenResult,
    ExpenseCategoryBreakdown,
    ExpenseCategoryCreate,
    ExpenseCategoryResponse,
    ExpenseCreate,
    ExpenseListResponse,
    ExpenseResponse,
    ExpenseUpdate,
    FCRCalculatorInput,
    FCRCalculatorResult,
    FeedNeedsInput,
    FeedNeedsResult,
    FinanceDashboardResponse,
    FinancialSnapshotResponse,
    ProfitProjectionInput,
    ProfitProjectionResult,
    RevenueListResponse,
    RevenueRecordCreate,
    RevenueRecordResponse,
    RevenueRecordUpdate,
)
from app.services import finance_service

router = APIRouter()


# ── Expense Categories ────────────────────────────────────────────────────────

@router.get(
    "/farms/{farm_id}/finance/categories",
    response_model=SuccessResponse[list[ExpenseCategoryResponse]],
    summary="List expense categories available to this farm",
    tags=["Finance"],
)
async def list_expense_categories(
    farm_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[list[ExpenseCategoryResponse]]:
    categories = await finance_service.list_expense_categories(db, farm_id)
    return SuccessResponse(data=categories)


@router.post(
    "/farms/{farm_id}/finance/categories",
    response_model=SuccessResponse[ExpenseCategoryResponse],
    status_code=status.HTTP_201_CREATED,
    summary="Create a custom expense category for this farm",
    tags=["Finance"],
)
async def create_expense_category(
    farm_id: UUID,
    body: ExpenseCategoryCreate,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[ExpenseCategoryResponse]:
    category = await finance_service.create_expense_category(db, farm_id, body, current_user)
    return SuccessResponse(data=category)


# ── Expenses ──────────────────────────────────────────────────────────────────

@router.post(
    "/farms/{farm_id}/expenses",
    response_model=SuccessResponse[ExpenseResponse],
    status_code=status.HTTP_201_CREATED,
    summary="Record a new expense",
    tags=["Finance"],
)
async def log_expense(
    farm_id: UUID,
    body: ExpenseCreate,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[ExpenseResponse]:
    expense = await finance_service.log_expense(db, farm_id, body, current_user)
    return SuccessResponse(data=expense)


@router.get(
    "/farms/{farm_id}/expenses",
    response_model=SuccessResponse[ExpenseListResponse],
    summary="List expenses for a farm",
    tags=["Finance"],
)
async def list_expenses(
    farm_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    flock_id: Optional[UUID] = Query(None),
    category_id: Optional[UUID] = Query(None),
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[ExpenseListResponse]:
    result = await finance_service.list_expenses(
        db, farm_id,
        flock_id=flock_id,
        category_id=category_id,
        date_from=date_from,
        date_to=date_to,
        page=page,
        page_size=page_size,
    )
    return SuccessResponse(data=result)


@router.get(
    "/farms/{farm_id}/expenses/{expense_id}",
    response_model=SuccessResponse[ExpenseResponse],
    summary="Get a single expense",
    tags=["Finance"],
)
async def get_expense(
    farm_id: UUID,
    expense_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[ExpenseResponse]:
    expense = await finance_service.get_expense(db, farm_id, expense_id)
    return SuccessResponse(data=expense)


@router.patch(
    "/farms/{farm_id}/expenses/{expense_id}",
    response_model=SuccessResponse[ExpenseResponse],
    summary="Correct an existing expense",
    tags=["Finance"],
)
async def update_expense(
    farm_id: UUID,
    expense_id: UUID,
    body: ExpenseUpdate,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[ExpenseResponse]:
    expense = await finance_service.update_expense(db, farm_id, expense_id, body, current_user)
    return SuccessResponse(data=expense)


@router.delete(
    "/farms/{farm_id}/expenses/{expense_id}",
    response_model=SuccessResponse[None],
    status_code=status.HTTP_200_OK,
    summary="Soft-delete an expense",
    tags=["Finance"],
)
async def delete_expense(
    farm_id: UUID,
    expense_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[None]:
    await finance_service.delete_expense(db, farm_id, expense_id)
    return SuccessResponse(data=None, message="Expense deleted")


# ── Revenue Records ───────────────────────────────────────────────────────────

@router.post(
    "/farms/{farm_id}/revenue",
    response_model=SuccessResponse[RevenueRecordResponse],
    status_code=status.HTTP_201_CREATED,
    summary="Record a new revenue event",
    tags=["Finance"],
)
async def log_revenue(
    farm_id: UUID,
    body: RevenueRecordCreate,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[RevenueRecordResponse]:
    record = await finance_service.log_revenue(db, farm_id, body, current_user)
    return SuccessResponse(data=record)


@router.get(
    "/farms/{farm_id}/revenue",
    response_model=SuccessResponse[RevenueListResponse],
    summary="List revenue records for a farm",
    tags=["Finance"],
)
async def list_revenue(
    farm_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    flock_id: Optional[UUID] = Query(None),
    revenue_type: Optional[str] = Query(None),
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[RevenueListResponse]:
    result = await finance_service.list_revenue(
        db, farm_id,
        flock_id=flock_id,
        revenue_type=revenue_type,
        date_from=date_from,
        date_to=date_to,
        page=page,
        page_size=page_size,
    )
    return SuccessResponse(data=result)


@router.get(
    "/farms/{farm_id}/revenue/{record_id}",
    response_model=SuccessResponse[RevenueRecordResponse],
    summary="Get a single revenue record",
    tags=["Finance"],
)
async def get_revenue_record(
    farm_id: UUID,
    record_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[RevenueRecordResponse]:
    record = await finance_service.get_revenue_record(db, farm_id, record_id)
    return SuccessResponse(data=record)


@router.patch(
    "/farms/{farm_id}/revenue/{record_id}",
    response_model=SuccessResponse[RevenueRecordResponse],
    summary="Correct a revenue record",
    tags=["Finance"],
)
async def update_revenue_record(
    farm_id: UUID,
    record_id: UUID,
    body: RevenueRecordUpdate,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[RevenueRecordResponse]:
    record = await finance_service.update_revenue_record(
        db, farm_id, record_id, body, current_user
    )
    return SuccessResponse(data=record)


@router.delete(
    "/farms/{farm_id}/revenue/{record_id}",
    response_model=SuccessResponse[None],
    status_code=status.HTTP_200_OK,
    summary="Soft-delete a revenue record",
    tags=["Finance"],
)
async def delete_revenue_record(
    farm_id: UUID,
    record_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[None]:
    await finance_service.delete_revenue_record(db, farm_id, record_id)
    return SuccessResponse(data=None, message="Revenue record deleted")


# ── Flock Financial Snapshot ──────────────────────────────────────────────────

@router.get(
    "/farms/{farm_id}/flocks/{flock_id}/finance",
    response_model=SuccessResponse[FinancialSnapshotResponse],
    summary="Get pre-computed P&L snapshot for a flock",
    tags=["Finance"],
)
async def get_flock_snapshot(
    farm_id: UUID,
    flock_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[FinancialSnapshotResponse]:
    snapshot = await finance_service.get_financial_snapshot(db, farm_id, flock_id)
    return SuccessResponse(data=snapshot)


@router.post(
    "/farms/{farm_id}/flocks/{flock_id}/finance/refresh",
    response_model=SuccessResponse[FinancialSnapshotResponse],
    summary="Force recompute the financial snapshot for a flock",
    tags=["Finance"],
)
async def refresh_flock_snapshot(
    farm_id: UUID,
    flock_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager"})),
    _perm=Depends(require_permission(Permission.FINANCE_RECORD)),
) -> SuccessResponse[FinancialSnapshotResponse]:
    snapshot = await finance_service.recompute_snapshot(db, farm_id, flock_id)
    from app.core.exceptions import NotFoundException
    if not snapshot:
        raise NotFoundException("Flock not found")
    from app.schemas.finance import FinancialSnapshotResponse
    return SuccessResponse(data=FinancialSnapshotResponse.model_validate(snapshot))


# ── Finance Dashboard ─────────────────────────────────────────────────────────

@router.get(
    "/farms/{farm_id}/finance",
    response_model=SuccessResponse[FinanceDashboardResponse],
    summary="Farm-level finance dashboard (reads from pre-computed snapshots)",
    tags=["Finance"],
)
async def get_finance_dashboard(
    farm_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[FinanceDashboardResponse]:
    farm, _ = access
    dashboard = await finance_service.get_finance_dashboard(db, farm, current_user)
    return SuccessResponse(data=dashboard)


@router.get(
    "/farms/{farm_id}/finance/categories/breakdown",
    response_model=SuccessResponse[list[ExpenseCategoryBreakdown]],
    summary="Expense breakdown by category for a date range",
    tags=["Finance"],
)
async def get_category_breakdown(
    farm_id: UUID,
    db: DBSession,
    current_user: CurrentUser,
    flock_id: Optional[UUID] = Query(None),
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    access: tuple = Depends(require_farm_access({"farm_owner", "farm_manager", "farm_worker", "vet_consultant", "viewer"})),
    _perm=Depends(require_permission(Permission.FINANCE_VIEW)),
) -> SuccessResponse[list[ExpenseCategoryBreakdown]]:
    breakdown = await finance_service.get_category_breakdown(
        db, farm_id,
        flock_id=flock_id,
        date_from=date_from,
        date_to=date_to,
    )
    return SuccessResponse(data=breakdown)


# ── Financial Calculators ─────────────────────────────────────────────────────

@router.post(
    "/calculators/fcr",
    response_model=SuccessResponse[FCRCalculatorResult],
    summary="Calculate Feed Conversion Ratio (FCR)",
    tags=["Finance", "Calculators"],
)
async def calculate_fcr(
    body: FCRCalculatorInput,
    current_user: CurrentUser,
) -> SuccessResponse[FCRCalculatorResult]:
    """Pure calculation — no DB access required. Available to any authenticated user."""
    result = finance_service.calculate_fcr(body)
    return SuccessResponse(data=result)


@router.post(
    "/calculators/profit-projection",
    response_model=SuccessResponse[ProfitProjectionResult],
    summary="Project flock profit at sale",
    tags=["Finance", "Calculators"],
)
async def calculate_profit_projection(
    body: ProfitProjectionInput,
    current_user: CurrentUser,
) -> SuccessResponse[ProfitProjectionResult]:
    result = finance_service.calculate_profit_projection(body)
    return SuccessResponse(data=result)


@router.post(
    "/calculators/break-even",
    response_model=SuccessResponse[BreakEvenResult],
    summary="Calculate break-even sale price",
    tags=["Finance", "Calculators"],
)
async def calculate_break_even(
    body: BreakEvenInput,
    current_user: CurrentUser,
) -> SuccessResponse[BreakEvenResult]:
    result = finance_service.calculate_break_even(body)
    return SuccessResponse(data=result)


@router.post(
    "/calculators/feed-needs",
    response_model=SuccessResponse[FeedNeedsResult],
    summary="Estimate feed needed to reach target weight",
    tags=["Finance", "Calculators"],
)
async def calculate_feed_needs(
    body: FeedNeedsInput,
    current_user: CurrentUser,
) -> SuccessResponse[FeedNeedsResult]:
    result = finance_service.calculate_feed_needs(body)
    return SuccessResponse(data=result)

File: frontend/src/api/finance.ts
/**
 * Greena — Finance API Client
 * All functions unwrap from APISuccess<T> envelope.
 * Covers: expense categories, expenses, revenue records,
 *         financial snapshots, dashboard, and calculators.
 */

import { apiClient } from "./client";
import type {
  APISuccess,
  BreakEvenResult,
  Expense,
  ExpenseCategory,
  ExpenseCategoryBreakdown,
  ExpenseCategoryCreateInput,
  ExpenseCreateInput,
  ExpenseListResponse,
  ExpenseUpdateInput,
  FCRCalculatorResult,
  FinanceDashboardResponse,
  FinancialSnapshot,
  FeedNeedsResult,
  ProfitProjectionResult,
  RevenueListResponse,
  RevenueRecord,
  RevenueRecordCreateInput,
  RevenueRecordUpdateInput,
} from "@/types";

// ── Expense Categories ────────────────────────────────────────────────────────

export async function listExpenseCategories(
  farmId: string
): Promise<ExpenseCategory[]> {
  const res = await apiClient.get<APISuccess<ExpenseCategory[]>>(
    `/farms/${farmId}/finance/categories`
  );
  return res.data.data;
}

export async function createExpenseCategory(
  farmId: string,
  body: ExpenseCategoryCreateInput
): Promise<ExpenseCategory> {
  const res = await apiClient.post<APISuccess<ExpenseCategory>>(
    `/farms/${farmId}/finance/categories`,
    body
  );
  return res.data.data;
}

// ── Expenses ──────────────────────────────────────────────────────────────────

export interface ExpenseListParams {
  flock_id?: string;
  category_id?: string;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
}

export async function logExpense(
  farmId: string,
  body: ExpenseCreateInput
): Promise<Expense> {
  const res = await apiClient.post<APISuccess<Expense>>(
    `/farms/${farmId}/expenses`,
    body
  );
  return res.data.data;
}

export async function listExpenses(
  farmId: string,
  params?: ExpenseListParams
): Promise<ExpenseListResponse> {
  const res = await apiClient.get<APISuccess<ExpenseListResponse>>(
    `/farms/${farmId}/expenses`,
    { params }
  );
  return res.data.data;
}

export async function getExpense(
  farmId: string,
  expenseId: string
): Promise<Expense> {
  const res = await apiClient.get<APISuccess<Expense>>(
    `/farms/${farmId}/expenses/${expenseId}`
  );
  return res.data.data;
}

export async function updateExpense(
  farmId: string,
  expenseId: string,
  body: ExpenseUpdateInput
): Promise<Expense> {
  const res = await apiClient.patch<APISuccess<Expense>>(
    `/farms/${farmId}/expenses/${expenseId}`,
    body
  );
  return res.data.data;
}

export async function deleteExpense(
  farmId: string,
  expenseId: string
): Promise<void> {
  await apiClient.delete(`/farms/${farmId}/expenses/${expenseId}`);
}

// ── Revenue Records ───────────────────────────────────────────────────────────

export interface RevenueListParams {
  flock_id?: string;
  revenue_type?: string;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
}

export async function logRevenue(
  farmId: string,
  body: RevenueRecordCreateInput
): Promise<RevenueRecord> {
  const res = await apiClient.post<APISuccess<RevenueRecord>>(
    `/farms/${farmId}/revenue`,
    body
  );
  return res.data.data;
}

export async function listRevenue(
  farmId: string,
  params?: RevenueListParams
): Promise<RevenueListResponse> {
  const res = await apiClient.get<APISuccess<RevenueListResponse>>(
    `/farms/${farmId}/revenue`,
    { params }
  );
  return res.data.data;
}

export async function getRevenueRecord(
  farmId: string,
  recordId: string
): Promise<RevenueRecord> {
  const res = await apiClient.get<APISuccess<RevenueRecord>>(
    `/farms/${farmId}/revenue/${recordId}`
  );
  return res.data.data;
}

export async function updateRevenueRecord(
  farmId: string,
  recordId: string,
  body: RevenueRecordUpdateInput
): Promise<RevenueRecord> {
  const res = await apiClient.patch<APISuccess<RevenueRecord>>(
    `/farms/${farmId}/revenue/${recordId}`,
    body
  );
  return res.data.data;
}

export async function deleteRevenueRecord(
  farmId: string,
  recordId: string
): Promise<void> {
  await apiClient.delete(`/farms/${farmId}/revenue/${recordId}`);
}

// ── Financial Snapshot ────────────────────────────────────────────────────────

export async function getFlockSnapshot(
  farmId: string,
  flockId: string
): Promise<FinancialSnapshot> {
  const res = await apiClient.get<APISuccess<FinancialSnapshot>>(
    `/farms/${farmId}/flocks/${flockId}/finance`
  );
  return res.data.data;
}

export async function refreshFlockSnapshot(
  farmId: string,
  flockId: string
): Promise<FinancialSnapshot> {
  const res = await apiClient.post<APISuccess<FinancialSnapshot>>(
    `/farms/${farmId}/flocks/${flockId}/finance/refresh`
  );
  return res.data.data;
}

// ── Dashboard ─────────────────────────────────────────────────────────────────

export async function getFinanceDashboard(
  farmId: string
): Promise<FinanceDashboardResponse> {
  const res = await apiClient.get<APISuccess<FinanceDashboardResponse>>(
    `/farms/${farmId}/finance`
  );
  return res.data.data;
}

export async function getCategoryBreakdown(
  farmId: string,
  params?: { flock_id?: string; date_from?: string; date_to?: string }
): Promise<ExpenseCategoryBreakdown[]> {
  const res = await apiClient.get<APISuccess<ExpenseCategoryBreakdown[]>>(
    `/farms/${farmId}/finance/categories/breakdown`,
    { params }
  );
  return res.data.data;
}

// ── Calculators ───────────────────────────────────────────────────────────────

export async function calculateFCR(
  total_feed_kg: number,
  total_live_weight_kg: number
): Promise<FCRCalculatorResult> {
  const res = await apiClient.post<APISuccess<FCRCalculatorResult>>(
    "/calculators/fcr",
    { total_feed_kg, total_live_weight_kg }
  );
  return res.data.data;
}

export async function calculateProfitProjection(body: {
  current_bird_count: number;
  expected_close_weight_kg: number;
  expected_sale_price_per_kg: number;
  total_expenses_so_far_kes: number;
  expected_additional_expenses_kes?: number;
  expected_mortality_pct?: number;
}): Promise<ProfitProjectionResult> {
  const res = await apiClient.post<APISuccess<ProfitProjectionResult>>(
    "/calculators/profit-projection",
    body
  );
  return res.data.data;
}

export async function calculateBreakEven(body: {
  total_expenses_kes: number;
  expected_birds_sold: number;
  expected_avg_weight_kg: number;
}): Promise<BreakEvenResult> {
  const res = await apiClient.post<APISuccess<BreakEvenResult>>(
    "/calculators/break-even",
    body
  );
  return res.data.data;
}

export async function calculateFeedNeeds(body: {
  current_bird_count: number;
  current_avg_weight_kg: number;
  target_weight_kg: number;
  target_fcr?: number;
  days_remaining?: number;
}): Promise<FeedNeedsResult> {
  const res = await apiClient.post<APISuccess<FeedNeedsResult>>(
    "/calculators/feed-needs",
    body
  );
  return res.data.data;
}
